uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php

```diff
@@ -23,29 +23,31 @@ public function upgrade_1018 () {
     $leaveRequestTable = LeaveRequest::getTableName();
     $absenceTypeTable = AbsenceType::getTableName();
 
-    $query = "SELECT leaveRequest.id
-      FROM {$leaveRequestTable} leaveRequest
-      INNER JOIN {$absenceTypeTable} absenceType ON leaveRequest.type_id = absenceType.id
-      WHERE absenceType.calculation_unit = {$calculationUnits['hours']}
-      AND leaveRequest.from_date IS NOT NULL
-      AND leaveRequest.from_date = leaveRequest.to_date
-      AND leaveRequest.is_deleted = 0";
-
-    $leaveRequestRecord = CRM_Core_DAO::executeQuery($query);
+    $query = "SELECT id FROM {$absenceTypeTable} WHERE calculation_unit = {$calculationUnits['hours']}";
+
+    $result = CRM_Core_DAO::executeQuery($query)->fetchAll();
+    $absenceTypesInHour = array_column($result, 'id');
+
+    $leaveRequest = new LeaveRequest();
+    $leaveRequest->whereAdd('from_date IS NOT NULL');
+    $leaveRequest->whereAdd('from_date = to_date');
+    $leaveRequest->whereAdd('is_deleted = 0');
+    $leaveRequest->whereAdd('type_id IN (' . implode(',', $absenceTypesInHour) . ')');
+    $number_of_rows = $leaveRequest->find();
+
     $contactWorkPatternService = new ContactWorkPatternService();
 
-    while ($leaveRequestRecord->fetch()) {
-      $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
+    while ($leaveRequest->fetch()) {
       $workDay = $contactWorkPatternService->getContactWorkDayForDate(
         $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));
 
       if (!empty($workDay['time_to'])) {
         $leaveRequest->to_date =
-          substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';
+          date("Y-m-d {$workDay['time_to']}:00", strtotime($leaveRequest->from_date));
 
-        $leaveRequest->save();
+        $leaveRequest->update();
```

Comment:
Why the change from `save` to `update`? `save` will call the `update()` method if the object has an ID, which I believe is the case here

Reply:
Save() will not work here due to calling of ->free().